From ca74b3a64d55e2d33d7b9ae8336f1e7bdc0ca8df Mon Sep 17 00:00:00 2001
From: Gabor Juhos <juhosg@openwrt.org>
Date: Thu, 16 Mar 2023 17:49:00 +0530
Subject: [PATCH 071/500] OpenWrt:
 703-phy-add-detach-callback-to-struct-phy_driver.patch

generic: add detach callback to struct phy_driver

lede-commit: fe61fc2d7d0b3fb348b502f68f98243b3ddf5867

Signed-off-by: Gabor Juhos <juhosg@openwrt.org>

Change-Id: Ib4df9c86073644ceaada4f308cf7f79390193427
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
(cherry picked from commit 88557f37f08a3caf4fdaab6662814358ccc95424)
---
 drivers/net/phy/phy_device.c | 3 +++
 include/linux/phy.h          | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/drivers/net/phy/phy_device.c b/drivers/net/phy/phy_device.c
index 2ce74593d6e4..c15d8e8c4cde 100644
--- a/drivers/net/phy/phy_device.c
+++ b/drivers/net/phy/phy_device.c
@@ -1797,6 +1797,9 @@ void phy_detach(struct phy_device *phydev)
 	struct module *ndev_owner = NULL;
 	struct mii_bus *bus;
 
+	if (phydev->drv && phydev->drv->detach)
+		phydev->drv->detach(phydev);
+
 	if (phydev->devlink)
 		device_link_del(phydev->devlink);
 
diff --git a/include/linux/phy.h b/include/linux/phy.h
index 1351b802ffcf..007fff29999c 100644
--- a/include/linux/phy.h
+++ b/include/linux/phy.h
@@ -963,6 +963,12 @@ struct phy_driver {
 	/** @handle_interrupt: Override default interrupt handling */
 	irqreturn_t (*handle_interrupt)(struct phy_device *phydev);
 
+	/*
+	 * Called before an ethernet device is detached
+	 * from the PHY.
+	 */
+	void (*detach)(struct phy_device *phydev);
+
 	/** @remove: Clears up any memory if needed */
 	void (*remove)(struct phy_device *phydev);
 
-- 
2.34.1

